README.md
# KillingFloor2-Ubuntu

Generates a functional Killing Floor 2 linux server.



CHANGELOG.md
# KillingFloor2-Ubuntu CHANGELOG

This file is used to list changes made in each version of the KillingFloor2-Ubuntu cookbook.

# 0.1.0

Initial release.

- change 0
- change 1



spec/unit/recipes/default_spec.rb
#
# Cookbook:: KillingFloor2-Ubuntu
# Spec:: default
#
# Copyright:: 2019, The Authors, All Rights Reserved.

require 'spec_helper'

describe 'KillingFloor2-Ubuntu::default' do
  context 'When all attributes are default, on Ubuntu 18.04' do
    # for a complete list of available platforms and versions see:
    # https://github.com/chefspec/fauxhai/blob/master/PLATFORMS.md
    platform 'ubuntu', '18.04'

    it 'converges successfully' do
      expect { chef_run }.to_not raise_error
    end
  end
end


test/integration/default/default_test.rb
# InSpec test for recipe KillingFloor2-Ubuntu::default

# The InSpec reference, with examples and extensive documentation, can be
# found at https://www.inspec.io/docs/reference/resources/

describe package("steamcmd") do
  it { should be_installed }
end

describe user 'steam' do
  it { should exist }
end

describe service('kf2server.service') do
  it { should be_installed }
  it { should be_enabled }
  it { should be_running }
end

describe port(7777) do
  it { should be_listening }
  its('protocols') { should include 'udp' }
end
describe port(27015) do
  it { should be_listening }
  its('protocols') { should include 'udp' }
end
describe port(8080) do
  it { should be_listening }
  its('protocols') { should include 'tcp' }
end
#describe port(20560) do
#  it { should be_listening }
#  its('protocols') { should include 'udp' }
#end

# Outbreak
# describe port(123) do
#  it { should be_listening }
#  its('protocols') { should include 'udp' }
#end


metadata.rb
name 'KillingFloor2-Ubuntu'
maintainer 'The Authors'
maintainer_email 'you@example.com'
license 'All Rights Reserved'
description 'Installs/Configures KillingFloor2-Ubuntu'
long_description 'Installs/Configures KillingFloor2-Ubuntu'
version '0.1.0'
chef_version '>= 14.0'
depends 'firewall'

# The `issues_url` points to the location where issues for this cookbook are
# tracked.  A `View Issues` link will be displayed on this cookbook's page when
# uploaded to a Supermarket.
#
# issues_url 'https://github.com/<insert_org_here>/KillingFloor2-Ubuntu/issues'

# The `source_url` points to the development repository for this cookbook.  A
# `View Source` link will be displayed on this cookbook's page when uploaded to
# a Supermarket.
#
# source_url 'https://github.com/<insert_org_here>/KillingFloor2-Ubuntu'


recipes/default.rb
#
# Cookbook:: KillingFloor2-Ubuntu
# Recipe:: default
#
# Copyright:: 2019, The Authors, All Rights Reserved.

## Install all prereqs/Update OS
apt_update

ip_list = Socket.ip_address_list

#Add DNS caching so SteamCMD is less shit
apt_package 'dnsmasq'
service 'dnsmasq' do
    action :restart
end

user 'steam' do
    manage_home true
    home '/home/steam'
    shell '/bin/false'
    password "$1$UKFlKYPt$kXT2BX8vnHrgMQuB4OMlo1"
    sensitive true
end

#SteamCMD is slow as shit otherwise
bash 'Adjust File Settings' do
    code <<-EOH
    echo "fs.file-max=100000" >> /etc/sysctl.conf && sysctl -p
    echo "* soft nofile 1000000" >> /etc/security/limits.conf
    echo "* hard nofile 1000000" >> /etc/security/limits.conf
    echo "session required pam_limits.so" >> /etc/pam.d/common-session
    EOH
end

bash 'Accept Steam EULA' do
    code <<-EOH
    echo steam steam/license note "" | sudo debconf-set-selections
    echo steam steam/question select "I AGREE" | sudo debconf-set-selections
    EOH
end

apt_package 'steamcmd' do
    options "-y"
    action :install
end

# Firewall Configurations
firewall_rule 'Game Port' do
    port      7777
    direction :in
    interface 'eth1'
    protocol  :udp
    command   :allow
end

firewall_rule 'Query Port' do
    port      27015
    direction :in
    interface 'eth1'
    protocol  :udp
    command   :allow
end

firewall_rule 'Steam Port' do
    port      20560
    direction :in
    interface 'eth1'
    protocol  :udp
    command   :allow
end

# open standard http port to tcp traffic only; insert as first rule
firewall_rule 'http' do
    port     80
    protocol :tcp
    interface 'eth1'
    position 1
    command   :allow
end

firewall_rule 'Web Admin' do
    port      8080
    direction :in
    interface 'eth1'
    protocol  :tcp
    command   :allow
end


firewall_rule 'ssh' do
    port     22
    command  :allow
end

# Script File for Kf2 Update
%w[ /home/steam/games /home/steam/games/killingfloor].each do |path|
    directory path do
        owner 'steam'
        group 'steam'
        mode '0755'
        recursive true
    end
end

cookbook_file '/home/steam/games/killingfloor/Install-KF2.sh' do
    source 'Install-KF2.sh'
    owner 'steam'
    group 'steam'
    mode '0755'
    action :create
end

## Login and install KF2
execute 'Install KF2' do
    command "steamcmd +runscript /home/steam/games/killingfloor/Install-KF2.sh"
    live_stream true
end

['LinuxServer-KFEngine.ini','LinuxServer-KFGame.ini','KFZedVarient.ini'].each do |file|
    cookbook_file "/home/steam/games/killingfloor/KFGame/Config/#{file}" do
        source "#{file}"
        mode "0755"
        owner "steam"
        group "steam"
        action :create
    end
end

execute 'chown' do
    command "chown -R steam /home/steam/games/killingfloor"
    user "root"
end

execute 'chgrp' do
    command "chgrp -R steam /home/steam/games/killingfloor"
    user "root"
end

systemd_unit 'kf2server.service' do
    content({Unit: {
            Description: 'kf2server',
            After: 'network.target',
            StartLimitIntervalSec: 0,
        },
        Service: {
            Type: 'simple',
            Restart: 'always',
            RestartSec: 10,
            User: 'steam',
            ExecStart: '/home/steam/games/killingfloor/Binaries/Win64/KFGameSteamServer.bin.x86_64 kf-burningparis?game=Zedternal.WMGameInfo_Endless?difficulty=0?Multihome=' + ip_list[2].ip_address,
        },
        Install: {
            WantedBy: 'multi-user.target'
        }})
    action [:create, :enable]
end

# We need to autogenerate KFWeb.ini because KF2 replaces it on shutdown
execute "Start & Stop Server" do
    command 'timeout 10s /home/steam/games/killingfloor/Binaries/Win64/KFGameSteamServer.bin.x86_64 kf-burningparis'
    user 'steam'
    ignore_failure true
end

ruby_block 'replace_line' do
    block do
        file = Chef::Util::FileEdit.new('/home/steam/games/killingfloor/KFGame/Config/KFWeb.ini')
        file.search_file_replace_line('bEnabled=false', 'bEnabled=true')
        file.write_file
    end
    notifies :start, "systemd_unit[kf2server.service]", :delayed
end